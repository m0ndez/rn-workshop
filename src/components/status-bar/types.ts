import { PropsWithChildren } from "react";
import { ViewStyle } from "react-native";

export type RootStatusBarProps = PropsWithChildren<{
  title: string;
  onPressBack: () => void;
  rightButtonComponent?: JSX.Element;
  styles?: {
    root?: ViewStyle;
    contentWrapper?: ViewStyle;
    pressBackButtonWrapper?: ViewStyle;
  };
}>;
