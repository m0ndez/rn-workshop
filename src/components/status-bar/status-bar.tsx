import { Surface } from "../surface";
import { ComponentPropsWithoutRef, FC, forwardRef } from "react";
import { StyleSheet, View, ViewStyle } from "react-native";
import { IconButton, Text } from "react-native-paper";
import { RootStatusBarProps } from "./types";

const baseStyles = StyleSheet.create({
  root: {
    gap: 8,
  },
  contentWrapper: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center",
  },
  pressBackButtonWrapper: {
    position: "absolute",
    left: 0,
  },
});

const RootStatusBar: FC<RootStatusBarProps> = (props) => {
  const { title, onPressBack, children, rightButtonComponent, styles } = props;

  const { contentWrapper, pressBackButtonWrapper, root } = baseStyles;

  return (
    <Surface style={[root, styles?.root]}>
      <View style={[contentWrapper, styles?.contentWrapper]}>
        <View style={[pressBackButtonWrapper, styles?.pressBackButtonWrapper]}>
          <IconButton size={24} icon={"chevron-left"} onPress={onPressBack} />
        </View>
        <Text variant="titleMedium">{title}</Text>
        <StatusBarRightButton>{rightButtonComponent}</StatusBarRightButton>
      </View>
      {children}
    </Surface>
  );
};

const StatusBarRightButton = forwardRef<
  View,
  ComponentPropsWithoutRef<typeof View>
>((props, ref) => {
  const { children, style, ...otherProps } = props;

  const baseStyle: ViewStyle = {
    position: "absolute",
    right: 0,
  };

  return (
    <View ref={ref} style={[baseStyle, style]} {...otherProps}>
      {children}
    </View>
  );
});

const StatusBarContent = forwardRef<
  View,
  ComponentPropsWithoutRef<typeof View>
>((props, ref) => {
  const { children, style, ...otherProps } = props;
  const baseStyle: ViewStyle = {
    justifyContent: "center",
    gap: 4,
  };

  return (
    <View ref={ref} style={[baseStyle, style]} {...otherProps}>
      {children}
    </View>
  );
});

export const StatusBar = Object.assign(RootStatusBar, {
  Content: StatusBarContent,
});
