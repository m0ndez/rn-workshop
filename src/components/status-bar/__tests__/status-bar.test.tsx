import React from "react";
import renderer from "react-test-renderer";
import { StatusBar } from "../status-bar";

describe("StatusBar", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<StatusBar title="Snapshot test!" onPressBack={() => {}} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
