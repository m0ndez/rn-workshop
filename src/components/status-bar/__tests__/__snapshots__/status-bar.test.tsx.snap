// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusBar renders correctly 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "rgb(247, 243, 249)",
      "borderRadius": 8,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 3,
    }
  }
  testID="surface-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgb(247, 243, 249)",
        "borderRadius": 8,
        "flex": undefined,
        "gap": 8,
        "padding": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 1,
      }
    }
    testID="surface"
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          [
            {
              "left": 0,
              "position": "absolute",
            },
            undefined,
          ]
        }
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "transparent",
              "borderRadius": 20,
              "height": 40,
              "margin": 6,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
              "width": 40,
            }
          }
          testID="icon-button-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "transparent",
                "borderColor": "rgba(121, 116, 126, 1)",
                "borderRadius": 20,
                "borderWidth": 0,
                "elevation": 0,
                "flex": 1,
                "overflow": "hidden",
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="icon-button-container"
          >
            <View
              accessibilityComponentType="button"
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityTraits="button"
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              centered={true}
              collapsable={false}
              focusable={true}
              hitSlop={
                {
                  "bottom": 6,
                  "left": 6,
                  "right": 6,
                  "top": 6,
                }
              }
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  [
                    {
                      "alignItems": "center",
                      "flexGrow": 1,
                      "justifyContent": "center",
                    },
                    {
                      "borderRadius": 20,
                    },
                  ],
                ]
              }
              testID="icon-button"
            >
              <Text />
            </View>
          </View>
        </View>
      </View>
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "500",
                "letterSpacing": 0.15,
                "lineHeight": 24,
              },
              undefined,
            ],
          ]
        }
      >
        Snapshot test!
      </Text>
      <View
        style={
          [
            {
              "position": "absolute",
              "right": 0,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;
